### DISCUSSION TITLE
Why Ocelot not AWS Gateway?

### DISCUSSION THREAD

**Discussion by ks1990cn** [ThreeMammals/Ocelot#1763] [Q&A] [answered]
Hi,

Can anyone explain what is perk of ocelot gateway over AWS gateway, just its free gateway? Is it safe as AWS gateway and much scalable and secure?

**Comment by raman-m** [comment]
Hi Tanmay!

Difficult question, without seeing AWS Gateway repository...
Is AWS Gateway open-source?
Is it coded in C# ? 😉  Do you have access to the AWS source code? I assume not.

**Comment by ks1990cn** [comment]
Amazon don't show their code!

I seached same question over Google Bard regarding Ocelot vs AWS Gateway API, you can go and is there any thing else want to add you can add, but till this time it gives me satisfactory answer.

Questions :- Can we think for adding monitoring implementation with Ocelot? Do we additional repo to implement this gateway with .Net code and hence we can route and manage like any other  API gateways.

It showed me following differences.

```
AWS API Gateway is a fully managed service, which means that Amazon takes care of all the provisioning, scaling, and maintenance. It is also highly scalable and can handle millions of requests per second. AWS API Gateway supports a wide range of features, including:

Authentication and authorization
Rate limiting
Caching
API versioning
Monitoring and analytics
```

```
Ocelot is an open-source API gateway that is built on .NET Core. It is lightweight and easy to use, but it does not offer as many features as AWS API Gateway. Ocelot supports the following features:

Routing
Authentication and authorization
Load balancing
Caching
Circuit breaker
```

<html>
<body>


Feature | AWS API Gateway | Ocelot
-- | -- | --
Managed service | Yes | No
Scalability | High | Medium
Features | Authentication, authorization, rate limiting, caching, API versioning, monitoring, analytics | Routing, authentication and authorization, load balancing, caching, circuit breakers
Platform | Cloud-agnostic | .NET Core
Open source | No | Yes


</body>
</html>

**Comment by RaynaldM** [comment]
Hi Tanmay,
It all depends on what you want to achieve.
If you want to be able to use the same gateway on AWS and Azure (or Google), you must not be tied to one of their technologies, and so a gateway like Ocelot is the obvious choice.
In our case, that's one of the reasons why we chose Ocelot, the other being the ability to "work" on requests in [Delegating Handlers](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html), and that gives us great flexibility.
The last point is that Ocelot is open-source and its code is easy to learn.

**Comment by ks1990cn** [comment]
I understand complete funda for ocelot. Was just unable to think when to use Ocelot and when to use AWS/ Azure gateway, what is exact use case?

If we host any website and use Ocelot, then also we will be paying cause any RestAPI will be using EC2 or Serverless system.

For open source and modification, its no doubt best option... I will be continue to learn on this and start contribution soon. Thanks

**Reply by raman-m** [reply]
Hi Tanmay! @ks1990cn 
What are your results in learning Ocelot?
I hope an achievable results, right?!
Does your development team still use Ocelot, or did they switch to AWS forever?

**Comment by raman-m** [comment]
@ks1990cn 
Dear Tanmay,

We don't care about AWS services! We care about Ocelot only!
We don't need any comparison tables because it is incorrect to compare paid service with closed code to an open-source project. **They cannot be compared at all!**
In theory only atomic features can be compared.

### First,
Scalability of Ocelot depends on hosting environment. Ocelot can be built as Docker image based on Kestrel web server, and after that Docker image can be instantiated as Docker container in environment with scalability options. For example in custom Kubernetes cluster, including cloud clusters like Azure Kubernetes Service and Amazon EKS.

### Second,
As I said, Ocelot is open source. We ensure some fairness and honesty when delivering Ocelot features. So, as an open-source product, Ocelot is open for extensions and audit. Can you audit Amazon service? No!

### Third, 
Ocelot is open for security audit. We don't track requests, we don't analyze request bodies to sniff the client's traffic. Ocelot has no integrations to some persistent storages to save statistics. Can you say the same about AWS Gateway?

### Fifth,
As a standard ASP.NET app, Ocelot can be deployed into private environments aka on-premises to serve the traffic of private networks of companies with high security standards, who care about data privacy and security. Can you say the same about Amazon Gateway?

### Finally,
As I said, that's wrong to compare closed-source products to open-source products. 

@RaynaldM  For having fun! FYI! 😉

**Comment by raman-m** [comment]
@ks1990cn 
> I will be continue to learn on this and start contribution soon.

We welcome your intention to return back to us with some interesting features/issues and PRs!
Good luck in learning Ocelot!

**Reply by raman-m** [reply]
@ks1990cn  Tanmay?

**Comment by raman-m** [comment, marked as answer]
@ks1990cn commented [on Sep 22](#discussioncomment-7421638)
> Feature | AWS API Gateway | Ocelot
> -- | -- | --
> Managed service | Yes | No

In what terms you understand "managed" here? Having some nice buttons to press on web-site? That's management, right?
As a developer, you can extend, build, and deploy Ocelot as you wish.

---
> Feature | AWS API Gateway | Ocelot
> -- | -- | --
> Scalability | High | Medium

I would say, Ocelot has high scalability also, because of building as Docker image. Scalability has provided by containerization services.

> Feature | AWS API Gateway | Ocelot
> -- | -- | --
> Features | Authentication, authorization, rate limiting, caching, API versioning, monitoring, analytics | Routing, authentication and authorization, load balancing, caching, circuit breakers

Every API gateway has its own nice features, and development teams choose what features they need to build a software product.

---
> Feature | AWS API Gateway | Ocelot
> -- | -- | --
> Platform | Cloud-agnostic | .NET Core

To be cloud-agnostic means AWS Gateway can be hosted in Azure and Google clouds... Hmmm... Really?
Could you share a link to prove that plz?
As .NET app, Ocelot can be deployed to **any** cloud too!

---
> Feature | AWS API Gateway | Ocelot
> -- | -- | --
> Open source | No | Yes

Totally agree!
You can pay for Amazon services. Jeff Bezos needs new one villa more in India 🤣

**Reply by raman-m** [reply]
### AI generated answer
- [Difference between Ocelot and AWS Gateway](https://gurubase.io/g/ocelot/difference-between-ocelot-and-aws-gateway)

**Comment by ks1990cn** [comment]
Thank you Raman for helping with nice and discriptive answers

**Reply by raman-m** [reply]
You are welcome!

**Reply by raman-m** [reply]
What I missed 3 years ago was your concern about Ocelot's monitoring capabilities — and you were absolutely right. AWS Gateway, Azure, and GCP all offer nice dashboards for tracking app state, traffic, service capacity, logging, and tracing.
Ocelot, however, provides no monitoring out of the box. That's a major disadvantage for our project, except maybe a Logging and Tracing feats, but they are not a monitor.
As project coordinator, I need to give this more thought...
